### ISSUE TITLE
log4j killing me, lang3 no better (related to issue 113)

### ISSUE THREAD

**Issue by DaveAccenture** [Azure/azure-iot-sdk-java#319]
Looks like I'm not the only one finding this issue: https://github.com/Azure/azure-iot-sdk-java/issues/313   I don't know why that's closed, it's still a big problem. I'm basically dead in the water.

I tried backing up to previous versions of this library. It seems before about 1.13.2 this was not the issue. BUT: It gave me a different error. SO, if you're not fixing issue 113, do you have a work around for this?:

Using    

` compile('com.microsoft.azure.sdk.iot:iot-device-client:1.13.0')
`
I get:

```
Program type already present: org.apache.commons.lang3.CharSetUtils
Message{kind=ERROR, text=Program type already present: org.apache.commons.lang3.CharSetUtils, sources=[Unknown source file], tool name=Optional.of(D8)}
```

I've dumped dependencies and don't see a problem with them. 

`org.apache.commons:commons-lang3:3.3.2`

is found several places, but all use the same version.

I really need this like yesterday, any help appreciated.

Using Android Studio 3.1.2
Windows 10
Java runtime 1.7, or 1.8 you tell me, which will work?
Android SDK : min 19, max 27, but I'll use any in that range that will work (this will only be run on two phones)

**Comment by JMayrbaeurl** [comment]
We changed from Log4j version 1 to version 2 with 1.13.2, because version 1 is legacy since 2015 (see #150 ) and using the Device Client SDK with modern frameworks like Spring or Spring Boot was not possible. But since we always wanted to remove the direct dependency on Log4j and enable the usage of other logging libraries like logback, we've finally replaced it with the usage of the logging facade implementation slf4j (see #310) on the current master branch. Please test with a build from the master branch.

**Comment by DaveAccenture** [comment]
Sounds promising. Sorry, I’m quite a novice at Java, can you please tell me how I would integrate this into my project?  I have the repo, but how do I modify my build file(s) or other file(s) to use my repo version rather than Maven’s cached version?

Dave Arendash

Tech Arch Science Manager
Immersive XR Experiences and VR Video SME
[cid:image001.png@01D35E01.ED605CD0]

**Comment by JMayrbaeurl** [comment]
I just build the Android sample application with the latest sources from the master branch (no longer using log4j2 directly, but using sl4j logging facade) and it compiles without errors. So you just have to wait some days (approx. till end of August) when the next version 1.13.4 will be released. Short notice. I had to increase the minimum SDK version from 19 to 21.

**Comment by DaveAccenture** [comment]
Is there any work around I can do today? My project depends on this working. Note thr sample I had used iot version 1.5.something and it worked fine. I don't mind using an older version as long as I can build it. Older version has com. apache. common. lang3 probems I have not yet found a solution to. I removed code that used it and attempted to exclude it, it still errs the same way, see above

**Comment by JMayrbaeurl** [comment]
I've posted a working build.gradle file in #313 for 1.13.1. Give it a try!

**Comment by DaveAccenture** [comment]
That helped a bit. I did find the problem. The bug is that Android Studio failed to find the lang3 dependency and file that were buried in a sub-project that had been eluding me, and that the gradle dependency dump did not list it either. I stumbled across the mismatched library version, replaced the local copy it was using with a maven reference, and it built ok.

Now, trying to get Azure working! See you on another thread....

**Comment by jasmineymlo** [comment]
Fixed in SDK v1.14.0.  Thanks @JMayrbaeurl 
--

**Comment by az-iot-builder-01** [comment]
@JMayrbaeurl, @DaveAccenture, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
